From 4d0854a5934456d57dc56fbed887e6e6b6866dd2 Mon Sep 17 00:00:00 2001
From: IzzelAliz <csh2001331@126.com>
Date: Sat, 19 Aug 2023 20:15:53 +0800
Subject: [PATCH] Add GetPlayerServer channel


diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index 11639f9c..a68c1538 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -33,6 +33,7 @@ import net.md_5.bungee.api.chat.BaseComponent;
 import net.md_5.bungee.api.chat.TextComponent;
 import net.md_5.bungee.api.config.ServerInfo;
 import net.md_5.bungee.api.connection.ProxiedPlayer;
+import net.md_5.bungee.api.connection.Server;
 import net.md_5.bungee.api.event.PluginMessageEvent;
 import net.md_5.bungee.api.event.ServerConnectEvent;
 import net.md_5.bungee.api.event.ServerDisconnectEvent;
@@ -427,6 +428,29 @@ public class DownstreamBridge extends PacketHandler
                     }
                     break;
                 }
+                // lightfall start
+                case "GetPlayerServer":
+                {
+                    String name = in.readUTF();
+                    ProxiedPlayer player = bungee.getPlayer( name );
+                    out.writeUTF( "GetPlayerServer" );
+                    out.writeUTF( name );
+                    if ( player == null )
+                    {
+                        out.writeUTF( "" );
+                        break;
+                    }
+                    Server srv = player.getServer();
+                    if ( srv == null )
+                    {
+                        out.writeUTF( "" );
+                    } else
+                    {
+                        out.writeUTF( srv.getInfo().getName() );
+                    }
+                    break;
+                }
+                // lightfall end
                 case "IP":
                     out.writeUTF( "IP" );
                     if ( con.getSocketAddress() instanceof InetSocketAddress )
-- 
2.36.1.windows.1

